dbms/src/Encryption/ReadBufferFromRandomAccessFileBuilder.cpp

```diff
@@ -13,23 +13,24 @@
 // limitations under the License.
 
 #include <Encryption/FileProvider.h>
-#include <Encryption/ReadBufferFromFileProvider.h>
+#include <Encryption/ReadBufferFromRandomAccessFileBuilder.h>
+
 
 namespace DB
 {
-ReadBufferFromFileProvider::ReadBufferFromFileProvider(
-    const FileProviderPtr & file_provider_,
+
+ReadBufferFromRandomAccessFilePtr ReadBufferFromRandomAccessFileBuilder::build(
```

Comment:
Why we return a Ptr rather an object here?

Reply:
maybe it can reduce a copy

Reply:
Could you give me an example?

Reply:
Meanwhile, will it bring about an extra indirect access, for example, we need to use `*it` rather than `it`. Will it change the performance if we use unique_ptr?

Reply:
I think the performance impact is small that we can just ignore.

Since `ReadBuffer` is a heavy class which usually own a large memory, copy & move are both expensive, I think use  pointer maybe a better idea.

Reply:
```cpp
ReadBufferFromRandomAccess ReadBufferFromRandomAccessFileBuilder::build() {
    return ReadBufferFromRandomAccess(xxx)
}

auto buf = ReadBufferFromRandomAccessFileBuilder::build(); // there is a copy, although it may be optimized.
```

Reply:
No, there are exactly no copy in your given example, even copies that you may think been optimized out.

Reply:
The C++ standard guarantees that no copy shall happen in this case, see https://en.cppreference.com/w/cpp/language/copy_elision

Reply:
yeah, I know this is be optimized since c++11. But we can not guarantees there will be no copy at all? Copy a unique_ptr cost much less than copy a ReadBuffer.

And as a lib function, return unique_ptr bring more convince. cast unique_ptr<ReadBuffer> to ReadBuffer is easy, but we can not cast ReadBuffer to unique_ptr<ReadBuffer> directly.

Reply:
> I know this is be optimized since c++11. But we can not guarantees there will be no copy at all?
Yes, we can. In fact the compiler MUST do all the RVO thing, since C++17.

> return unique_ptr bring more convince
Then return a not_null to reduce calling `if (xxx)` over and over again. Because it is hard to get the idea which unique_ptr is guaranteed to be not null.

Reply:
Meanwhile, `ReadBuffer` will create a vtable, since it has some virtual functions. I wonder if it's better to avoid use any kind of pointer, if we know exactly what we are operate on?

Reply:
Please check https://stackoverflow.com/questions/77925442/when-does-the-c-standard-permits-forces-the-compiler-to-devirtualization for the "use o->foo() rather than o.foo()" problem.